packages/oruga/src/composables/useInputHandler.ts

```diff
@@ -224,13 +229,81 @@ export function useInputHandler<T extends ValidatableFormElement>(
         emits("invalid", event);
     }
 
+    /**
+     * Provides a way to force the watcher on `updateCustomValidationMessage` to re-run
+     *
+     * There are some cases (e.g. changes to the element's validation attributes) that can
+     * force changes to the element's `validityState`, which isn't a reactive property.
+     * We call `updateCustomValidationMessage` in cases where we expect relevant changes
+     * to `validityState`. That's enough to update the DOM correctly. However, changes in
+     * `validityState` are likely to cause the function version of `props.customValidity`
+     * to execute different branches that may have different dependencies on reactive
+     * properties. The `watchEffect` call on `updateCustomValidationMessage` needs to know
+     * about those new reactive dependencies, and the only way to inform it of them is to
+     * re-run `updateCustomValidationMessage` within the watcher's context.
+     *
+     * This does mean that if you need to update the DOM element's validity state _immediately_,
+     * you'll need to call updateCustomValidationMessage directly (to make the change happen now
+     * rather than waiting for the reactivity system) _and_ trigger this `ref` (to register
+     * reactive dependencies), so `updateCustomValidationMessage` will run twice for those updates.
+     */
+    const forceValidationReactivityUpdate = ref(null);
```

Comment:
I don't think we need that here. You are thinking of something too complicated. 

The `watchEffect` calls a function `updateCustomValidationMessage`. 
What you are trying to do here is force the watcher to call this function via a trigger (`forceValidationReactivityUpdate`), then you want to call the trigger with `triggerRef(forceValidationReactivityUpdate);` by yourself.

But why doesn't you just call the function itself? It shouldn't make any difference whether the function is called by the watcher whenever a reactive dependency has changed, or by calling it by hand.

Reply:
> But why doesn't you just call the function itself? It shouldn't make any difference whether the function is called by the watcher whenever a reactive dependency has changed, or by calling it by hand.

The difference is that Vue only detects new reactive dependencies (e.g. ones that are in a branch that hadn't been called before) if the function is called inside the watcher's context. Before we added a reactive dependency on `props.modeValue`, this code had all sorts of reactivity issues where changes to an unrecognized reactive dependency would make the message stale. Adding that dependency has helped dramatically because changes in the control's value are _usually_ the main thing affecting `element.validityState`, but if (1) there's a branch where the validity state doesn't depend on the model value and (2) that branch is reached by changes in _validation attributes_ rather than the model value, things fall apart.

```vue
<script setup lang="ts">
import { computed, ref } from 'vue';

const val = ref('');
const required = ref(true);
const messageSwitch = ref(false);

const customMessage = computed(() => messageSwitch.value ? 'message B' : 'message A');

function validation(state: ValidityState): string {
  if (state.valueMissing) {
    return 'Enter a value.';
  } else {
    return customMessage.value;
  }
}

const onSubmit = (e: Event) => { alert('Submitted'); e.preventDefault(); };
</script>

<template>
  <form @submit="onSubmit">
    <OField label="Value">
      <OInput v-model="val" :required :customValidity="validation" />
    </OField>
    <OField>
      <OCheckbox v-model="required">Required</OCheckbox>
    </OField>
    <OField>
      <OCheckbox v-model="messageSwitch">Alternate message</OCheckbox>
    </OField>
    <OButton native-type="submit">Submit</OButton>
  </form>
</template>
```

Without `forceValidationReactivityUpdate`, if you blur the input box, you'll get the "Enter a value" message. If you then uncheck "required", you'll get "message A", but toggling the other checkbox doesn't start displaying "message B" like it should.

Reply:
I pushed up a fix for another issue I found while testing that example. I also found that the fix allowed for a refactoring that simplifies this logic a bit. The given example would still be broken without forcing the watcher to run again, though.